### PR TITLE
Add disable for Query validation

### LINKED ISSUE #4: GetQueryPoolResults in core validation is slow
**Issue by [karl-lunarg](https://github.com/karl-lunarg) (MIGRATED)**
_Friday Aug 05, 2016 at 20:24 GMT_
_Originally opened as https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/829_

----

Just logging this as the result of some analysis from LunarXchange issue 305.

Some perf analysis shows that GetQueryPoolResults spends about 80% of its time creating the temporary unordered_map for very large query pools (e.g., thousands).  Yes, this isn't an issue for smaller pools and creating the map certainly speeds up the subsequent validation operations.  But this overhead can be too costly for some applications.  (See LX 305 issue).  We wonder if there is a way to avoid creating this map.  Also, this general problem might be addressed in a future layers architecture where expensive tests can be disabled.

The issue can be reproduced fairly easily by modifying the occlusion_query sample app from the LunarG/VulkanSamples repo.  Change the queryCount argument in vkCmdResetQueryPool to a large number like 16,000.  (This is a missing validation check, btw) And comment out the call to vkQueueWaitIdle so that the CB with the queries is still in flight.  The subsequent call to vkGetQueryPoolResults then runs slowly.

Note also that the implementation of std::unordered_map in VS 2013 has serious performance issues, which made the (implicit) free of the unordered_map take MUCH longer than desired.  See http://stackoverflow.com/questions/21014822/very-slow-unordered-map-clearing.  Moving to VS 2015 changed the cost of the free from 850 ms on VS 2013 to 7 ms on VS 2015 for a particular test.  Since unordered_map is used a lot in the layers, this implies that using VS 2015 might help the perf of the VLs in a more general way.

### PR DESCRIPTION
**Pull request by mark-lunarg**
Added a disable for query validation. To disable, set this in your vk_layer_settings.txt file:

     `khronos.validation.disables = VALIDATION_CHECK_DISABLE_QUERY_VALIDATION`

-or- 

set the `VK_LAYER_DISABLES ` environment variable to `VALIDATION_CHECK_DISABLE_QUERY_VALIDATION`

Fixes #4.  Or, at least works around it.

The `VK_EXT_validation_features `extension will be modified to include a flag for this disable.

### CODE REVIEW

File: scripts/layer_chassis_generator.py [hunk 1]
```diff
@@ -275,6 +276,7 @@ class LAYER_PHYS_DEV_PROPERTIES {
     bool object_in_use;                             // Skip all object in_use checking
     bool idle_descriptor_set;                       // Skip check to verify that descriptor set is not in-use
     bool push_constant_range;                       // Skip push constant range checks
+    bool all_query_validation;                      // Disable all core validation query-related checks
```
Comment (tobine) [inline scripts/layer_chassis_generator.py]: Is the "all_" prefix redundant? Just "query_validation" seems sufficient to me.

Comment (mark-lunarg) [reply scripts/layer_chassis_generator.py]: Was trying to signify that this turned of the validation for all query-related APIs. Some of the other names are under-described, but will remove the 'all_'.

**Comment by mark-lunarg** [comment]
- The DynViewportParamTests failure is unrelated and happens on Master, I'll fix it separately.
 - A test cannot be added until the ValFeatures extension is updated with this disable.  Wrote one, but there's no backdoor between the LVTs and the layer to change the disables other than the ext.
 - I fairly strongly believe that these disables should only be for validation checks and that all state tracking should happen regardless.  Sprinkling the disable checks throughout the state tracking code would make things messy/complicated, and who knows what dependencies there are now, or _in the future_, where other validation code might need the query state.

**Comment by mark-lunarg** [comment]
@tobine, can you take one more look?

**Pull request merged**